framework/src/timesteppers/CompositionDT.C

```diff
@@ -0,0 +1,178 @@
+    _has_initial_dt(isParamValid("dt")),
+    _initial_dt(_has_initial_dt ? getParam<Real>("dt") : 0.),
+    _base_timestepper(getParam<std::string>("base_timestepper"))
+{
+  if (isParamValid("input_timesteppers"))
+    _inputs = parameters.get<std::vector<std::string>>("input_timesteppers");
+
+  if (isParamValid("times_to_hit_timestepper"))
+    _hit_timestepper_name = parameters.get<std::string>("times_to_hit_timestepper");
+}
+
+Real
+CompositionDT::computeInitialDT()
+{
+  if (_has_initial_dt)
+    return _initial_dt;
+  else
+    return computeDT();
+}
+
+Real
+CompositionDT::computeDT()
+{
+  std::map<const std::string, Real> dts;
+
+  for (const auto & name : _inputs)
+  {
+    auto time_stepper = setupTimeStepper(name);
+    Real dt = time_stepper->getCurrentDT();
+
+    dts.emplace(name, dt);
+  }
+
+  auto base_stepper = setupTimeStepper(_base_timestepper);
+  _dt = produceCompositionDT(std::as_const(dts), base_stepper->getCurrentDT());
+
+  return _dt;
+}
+
+std::shared_ptr<TimeStepper>
+CompositionDT::setupTimeStepper(const std::string & stpper_name)
+{
+  auto time_stepper = _app.getTimeStepperSystem().getTimeStepper(stpper_name);
+  mooseAssert(time_stepper, "Not exist");
+  time_stepper->computeStep();
+
+  return time_stepper;
+}
+
+std::shared_ptr<TimeSequenceStepperBase>
+CompositionDT::setupSequenceStepper(const std::string & stpper_name)
+{
+  auto time_stepper = _app.getTimeStepperSystem().getTimeStepper(stpper_name);
+  auto hit_timestepper = std::dynamic_pointer_cast<TimeSequenceStepperBase>(time_stepper);
+  if (!hit_timestepper)
+    mooseError("A TimeSequenceStepper type is required to generate times to hit");
+
+  hit_timestepper->init();
+
+  return hit_timestepper;
+}
+
+Real
+CompositionDT::maxTimeStep(const std::map<const std::string, Real> & dts)
+{
+  auto maxDT = std::max_element(
+      dts.begin(), dts.end(), [](const auto & x, const auto & y) { return x.second < y.second; });
+
+  return maxDT->second;
+}
+
+Real
+CompositionDT::minTimeStep(const std::map<const std::string, Real> & dts)
+{
+  auto minDT = std::min_element(
+      dts.begin(), dts.end(), [](const auto & a, const auto & b) { return a.second < b.second; });
+  return minDT->second;
+}
+
+Real
+CompositionDT::produceCompositionDT(const std::map<const std::string, Real> & dts,
+                                    const Real & base_dt)
+{
+  Real composeDT = 0;
+
+  auto composition_type = getParam<MooseEnum>("composition_type");
+
+  if (composition_type == "max")
+    composeDT = std::min(maxTimeStep(dts), base_dt);
+
+  else if (composition_type == "min")
+    composeDT = std::max(minTimeStep(dts), base_dt);
+
+  if (isParamValid("times_to_hit_timestepper"))
+  {
+    if (composeDT != 0)
+      return produceHitDT(composeDT);
+    else
+      return produceHitDT(base_dt);
+  }
+
+  return composeDT;
+}
+
+Real
+CompositionDT::produceHitDT(const Real & composeDT)
+{
+  auto ts = setupSequenceStepper(_hit_timestepper_name);
```

Comment:
this inits on every time step. that would be surprising?

Reply:
I think the function name is misleading. It is not init but get the current TimeSequence time.